ICE: cannot find .. in param-env: ParamEnv

<!--
[31mICE[0m: Rustc ./87DF6373DF3D4D8E3161997F661C265A8B21B672BCEEDCFA5D8B8ED787B8393D.rs '-Zincremental-verify-ich=yes -Cincremental=<dir> -Cdebuginfo=2 -Clink-dead-code=true -Zvalidate-mir -Zquery-dep-graph -Zthreads=4 --edition=2024' 'error: internal compiler error: compiler/rustc_middle/src/ty/sty.rs:373:13: cannot find `N/#0` in param-env: ParamEnv {', 'error: internal compiler error: compiler/rustc_middle/src/ty/sty.rs:373:13: cannot find `N/#0` in param-env: ParamEnv {'
File: /tmp/im/87DF6373DF3D4D8E3161997F661C265A8B21B672BCEEDCFA5D8B8ED787B8393D.rs
-->

unlike #143117, this one does neither require any feature gates nor compiler flags, it also repros on stable already

auto-reduced (treereduce-rust):
````rust

impl<
        const N: usize = {
            static || {
                Foo([0; X]);
            }
        },
    > PartialEq for True
{
}

````

original:
````rust


impl<
        const N: usize = {
            static || {
                Foo([0; X]);
            }
        },
    > PartialEq for True
{
}

````

Version information
````
rustc 1.92.0-nightly (f43597208 2025-10-24)
binary: rustc
commit-hash: f435972085b697a1ece8ee6a1ac76efff8d1df7b
commit-date: 2025-10-24
host: x86_64-unknown-linux-gnu
release: 1.92.0-nightly
LLVM version: 21.1.3

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/f435972085b697a1ece8ee6a1ac76efff8d1df7b/compiler/rustc_middle/src/ty/sty.rs#L367-L379

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zincremental-verify-ich=yes -Cincremental=<dir> -Cdebuginfo=2 -Clink-dead-code=true -Zvalidate-mir -Zquery-dep-graph -Zthreads=4 --edition=2024`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0412]: cannot find type `True` in this scope
 --> /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:7:21
  |
7 |     > PartialEq for True
  |                     ^^^^ not found in this scope
  |
help: you may want to use a bool value instead
  |
7 -     > PartialEq for True
7 +     > PartialEq for true
  |

error[E0425]: cannot find value `X` in this scope
 --> /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:4:25
  |
2 |         const N: usize = {
  |               - similarly named const parameter `N` defined here
3 |             static || {
4 |                 Foo([0; X]);
  |                         ^ help: a const parameter with a similar name exists: `N`

error[E0658]: coroutine syntax is experimental
 --> /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:3:13
  |
3 |             static || {
  |             ^^^^^^
  |
  = note: see issue #43122 <https://github.com/rust-lang/rust/issues/43122> for more information
  = help: add `#![feature(coroutines)]` to the crate attributes to enable
  = note: this compiler was built on 2025-10-24; consider upgrading it if it is out of date

error[E0697]: closures cannot be static
 --> /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:3:13
  |
3 |             static || {
  |             ^^^^^^^^^

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:9:2
  |
9 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs`

error: defaults for generic parameters are not allowed here
 --> /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:2:9
  |
2 | /         const N: usize = {
3 | |             static || {
4 | |                 Foo([0; X]);
5 | |             }
6 | |         },
  | |_________^

error: internal compiler error: compiler/rustc_middle/src/ty/sty.rs:373:13: cannot find `N/#0` in param-env: ParamEnv {
                                    caller_bounds: [],
                                }


thread 'rustc' (4015949) panicked at compiler/rustc_middle/src/ty/sty.rs:373:13:
Box<dyn Any>
stack backtrace:
   0:     0x7f486b634073 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h492c738ad49cea7d
   1:     0x7f486be01b98 - core::fmt::write::h1c21b4db123ae31c
   2:     0x7f486b5e8403 - std::io::Write::write_fmt::h78003e47d8f6c72e
   3:     0x7f486b5f9bf2 - std::sys::backtrace::BacktraceLock::print::h0d646c4d1cbf366e
   4:     0x7f486b5ffcc9 - std::panicking::default_hook::{{closure}}::h274ab31bbd3c2851
   5:     0x7f486b5ff7f3 - std::panicking::default_hook::he590f8320f1e009e
   6:     0x7f486a600611 - std[3e186dc2068ac4b9]::panicking::update_hook::<alloc[47fc8f5871530046]::boxed::Box<rustc_driver_impl[b652d2bf07caa0d8]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7f486b6000ef - std::panicking::panic_with_hook::hc0910476a8aaf08a
   8:     0x7f486a639e61 - std[3e186dc2068ac4b9]::panicking::begin_panic::<rustc_errors[345ffcadd731d51f]::ExplicitBug>::{closure#0}
   9:     0x7f486a62e056 - std[3e186dc2068ac4b9]::sys::backtrace::__rust_end_short_backtrace::<std[3e186dc2068ac4b9]::panicking::begin_panic<rustc_errors[345ffcadd731d51f]::ExplicitBug>::{closure#0}, !>
  10:     0x7f486a62b939 - std[3e186dc2068ac4b9]::panicking::begin_panic::<rustc_errors[345ffcadd731d51f]::ExplicitBug>
  11:     0x7f486a6459d1 - <rustc_errors[345ffcadd731d51f]::diagnostic::BugAbort as rustc_errors[345ffcadd731d51f]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:     0x7f486abdd4ea - rustc_middle[df80570e3f8f0476]::util::bug::opt_span_bug_fmt::<rustc_span[b6aef8613cdfc213]::span_encoding::Span>::{closure#0}
  13:     0x7f486abdd7ca - rustc_middle[df80570e3f8f0476]::ty::context::tls::with_opt::<rustc_middle[df80570e3f8f0476]::util::bug::opt_span_bug_fmt<rustc_span[b6aef8613cdfc213]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  14:     0x7f486abce53b - rustc_middle[df80570e3f8f0476]::ty::context::tls::with_context_opt::<rustc_middle[df80570e3f8f0476]::ty::context::tls::with_opt<rustc_middle[df80570e3f8f0476]::util::bug::opt_span_bug_fmt<rustc_span[b6aef8613cdfc213]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  15:     0x7f4868946320 - rustc_middle[df80570e3f8f0476]::util::bug::bug_fmt
  16:     0x7f486d913767 - <rustc_middle[df80570e3f8f0476]::ty::sty::ParamConst>::find_const_ty_from_env.cold
  17:     0x7f486c6e8e75 - <rustc_trait_selection[cf51c7a433a87fce]::traits::fulfill::FulfillProcessor as rustc_data_structures[9e234764dc4c351e]::obligation_forest::ObligationProcessor>::process_obligation
  18:     0x7f486be04e09 - <rustc_data_structures[9e234764dc4c351e]::obligation_forest::ObligationForest<rustc_trait_selection[cf51c7a433a87fce]::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection[cf51c7a433a87fce]::traits::fulfill::FulfillProcessor>
  19:     0x7f486cc1a6d9 - <rustc_hir_typeck[ca3afe1245f8583a]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  20:     0x7f486cc25059 - <rustc_hir_typeck[ca3afe1245f8583a]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  21:     0x7f486c0acdb7 - <rustc_hir_typeck[ca3afe1245f8583a]::fn_ctxt::FnCtxt>::check_expr_call
  22:     0x7f486cc1a994 - <rustc_hir_typeck[ca3afe1245f8583a]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  23:     0x7f486cc15903 - <rustc_hir_typeck[ca3afe1245f8583a]::fn_ctxt::FnCtxt>::check_expr_block
  24:     0x7f486cc1b097 - <rustc_hir_typeck[ca3afe1245f8583a]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  25:     0x7f486cc0db5a - rustc_hir_typeck[ca3afe1245f8583a]::check::check_fn
  26:     0x7f486c659787 - <rustc_hir_typeck[ca3afe1245f8583a]::fn_ctxt::FnCtxt>::check_expr_closure
  27:     0x7f486cc2231d - <rustc_hir_typeck[ca3afe1245f8583a]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  28:     0x7f486cc14b22 - <rustc_hir_typeck[ca3afe1245f8583a]::fn_ctxt::FnCtxt>::check_expr_block
  29:     0x7f486cc1b097 - <rustc_hir_typeck[ca3afe1245f8583a]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  30:     0x7f486c8faf7e - rustc_hir_typeck[ca3afe1245f8583a]::typeck_with_inspect::{closure#0}
  31:     0x7f486c8f6b30 - rustc_query_impl[ce892b7de41d33cc]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[ce892b7de41d33cc]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[df80570e3f8f0476]::query::erase::Erased<[u8; 8usize]>>
  32:     0x7f486bef472c - rustc_query_system[c36681704e76fc2e]::query::plumbing::try_execute_query::<rustc_query_impl[ce892b7de41d33cc]::DynamicConfig<rustc_data_structures[9e234764dc4c351e]::vec_cache::VecCache<rustc_span[b6aef8613cdfc213]::def_id::LocalDefId, rustc_middle[df80570e3f8f0476]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[c36681704e76fc2e]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[ce892b7de41d33cc]::plumbing::QueryCtxt, true>
  33:     0x7f486bef1b35 - rustc_query_impl[ce892b7de41d33cc]::query_impl::typeck::get_query_incr::__rust_end_short_backtrace
  34:     0x7f486c8d6ff3 - rustc_query_impl[ce892b7de41d33cc]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[ce892b7de41d33cc]::query_impl::used_trait_imports::dynamic_query::{closure#2}::{closure#0}, rustc_middle[df80570e3f8f0476]::query::erase::Erased<[u8; 8usize]>>
  35:     0x7f486bef472c - rustc_query_system[c36681704e76fc2e]::query::plumbing::try_execute_query::<rustc_query_impl[ce892b7de41d33cc]::DynamicConfig<rustc_data_structures[9e234764dc4c351e]::vec_cache::VecCache<rustc_span[b6aef8613cdfc213]::def_id::LocalDefId, rustc_middle[df80570e3f8f0476]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[c36681704e76fc2e]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[ce892b7de41d33cc]::plumbing::QueryCtxt, true>
  36:     0x7f486bef3c83 - rustc_query_impl[ce892b7de41d33cc]::query_impl::used_trait_imports::get_query_incr::__rust_end_short_backtrace
  37:     0x7f486c8d54cd - rustc_hir_analysis[83592e86f71d95cd]::check_unused::check_unused_traits
  38:     0x7f486c8d53cd - rustc_query_impl[ce892b7de41d33cc]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[ce892b7de41d33cc]::query_impl::check_unused_traits::dynamic_query::{closure#2}::{closure#0}, rustc_middle[df80570e3f8f0476]::query::erase::Erased<[u8; 0usize]>>
  39:     0x7f486d208e78 - rustc_query_system[c36681704e76fc2e]::query::plumbing::try_execute_query::<rustc_query_impl[ce892b7de41d33cc]::DynamicConfig<rustc_query_system[c36681704e76fc2e]::query::caches::SingleCache<rustc_middle[df80570e3f8f0476]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[ce892b7de41d33cc]::plumbing::QueryCtxt, true>
  40:     0x7f486d208723 - rustc_query_impl[ce892b7de41d33cc]::query_impl::check_unused_traits::get_query_incr::__rust_end_short_backtrace
  41:     0x7f486c8a590c - rustc_hir_analysis[83592e86f71d95cd]::check_crate
  42:     0x7f486c8cf770 - rustc_interface[10f842ee2745836b]::passes::analysis
  43:     0x7f486c8cf435 - rustc_query_impl[ce892b7de41d33cc]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[ce892b7de41d33cc]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[df80570e3f8f0476]::query::erase::Erased<[u8; 0usize]>>
  44:     0x7f486d208fc8 - rustc_query_system[c36681704e76fc2e]::query::plumbing::try_execute_query::<rustc_query_impl[ce892b7de41d33cc]::DynamicConfig<rustc_query_system[c36681704e76fc2e]::query::caches::SingleCache<rustc_middle[df80570e3f8f0476]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[ce892b7de41d33cc]::plumbing::QueryCtxt, true>
  45:     0x7f486d20883f - rustc_query_impl[ce892b7de41d33cc]::query_impl::analysis::get_query_incr::__rust_end_short_backtrace
  46:     0x7f486d13ec8a - <rustc_interface[10f842ee2745836b]::passes::create_and_enter_global_ctxt<core[e8997f4232dfd718]::option::Option<rustc_interface[10f842ee2745836b]::queries::Linker>, rustc_driver_impl[b652d2bf07caa0d8]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[e8997f4232dfd718]::ops::function::FnOnce<(&rustc_session[4dec040bcdb04b8e]::session::Session, rustc_middle[df80570e3f8f0476]::ty::context::CurrentGcx, alloc[47fc8f5871530046]::sync::Arc<rustc_data_structures[9e234764dc4c351e]::jobserver::Proxy>, &std[3e186dc2068ac4b9]::sync::once_lock::OnceLock<rustc_middle[df80570e3f8f0476]::ty::context::GlobalCtxt>, &rustc_data_structures[9e234764dc4c351e]::sync::worker_local::WorkerLocal<rustc_middle[df80570e3f8f0476]::arena::Arena>, &rustc_data_structures[9e234764dc4c351e]::sync::worker_local::WorkerLocal<rustc_hir[739a0293ab2ba12c]::Arena>, rustc_driver_impl[b652d2bf07caa0d8]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  47:     0x7f486cfb37e2 - rustc_interface[10f842ee2745836b]::interface::run_compiler::<(), rustc_driver_impl[b652d2bf07caa0d8]::run_compiler::{closure#0}>::{closure#1}
  48:     0x7f486a618734 - <rustc_thread_pool[7fc909dca122daad]::job::StackJob<rustc_thread_pool[7fc909dca122daad]::latch::LatchRef<rustc_thread_pool[7fc909dca122daad]::latch::LockLatch>, <rustc_thread_pool[7fc909dca122daad]::registry::Registry>::in_worker_cold<<rustc_thread_pool[7fc909dca122daad]::thread_pool::ThreadPool>::install<rustc_interface[10f842ee2745836b]::util::run_in_thread_pool_with_globals<rustc_interface[10f842ee2745836b]::interface::run_compiler<(), rustc_driver_impl[b652d2bf07caa0d8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#1}::{closure#0}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()> as rustc_thread_pool[7fc909dca122daad]::job::Job>::execute
  49:     0x7f486b30bb1f - <rustc_thread_pool[7fc909dca122daad]::registry::WorkerThread>::wait_or_steal_until_cold
  50:     0x7f486b30a5c1 - <rustc_thread_pool[7fc909dca122daad]::registry::ThreadBuilder>::run
  51:     0x7f486a5f803d - std[3e186dc2068ac4b9]::sys::backtrace::__rust_begin_short_backtrace::<<rustc_thread_pool[7fc909dca122daad]::ThreadPoolBuilder>::build_scoped<rustc_interface[10f842ee2745836b]::util::run_in_thread_pool_with_globals<rustc_interface[10f842ee2745836b]::interface::run_compiler<(), rustc_driver_impl[b652d2bf07caa0d8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#0}, rustc_interface[10f842ee2745836b]::util::run_in_thread_pool_with_globals<rustc_interface[10f842ee2745836b]::interface::run_compiler<(), rustc_driver_impl[b652d2bf07caa0d8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#1}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
  52:     0x7f486a603632 - <<std[3e186dc2068ac4b9]::thread::Builder>::spawn_unchecked_<<rustc_thread_pool[7fc909dca122daad]::ThreadPoolBuilder>::build_scoped<rustc_interface[10f842ee2745836b]::util::run_in_thread_pool_with_globals<rustc_interface[10f842ee2745836b]::interface::run_compiler<(), rustc_driver_impl[b652d2bf07caa0d8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#0}, rustc_interface[10f842ee2745836b]::util::run_in_thread_pool_with_globals<rustc_interface[10f842ee2745836b]::interface::run_compiler<(), rustc_driver_impl[b652d2bf07caa0d8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#1}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>::{closure#1} as core[e8997f4232dfd718]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  53:     0x7f486cbbf02f - std::sys::thread::unix::Thread::new::thread_start::h69c86f976e686da7
  54:     0x7f48666969cb - <unknown>
  55:     0x7f486671aa0c - <unknown>
  56:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.92.0-nightly (f43597208 2025-10-24) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z incremental-verify-ich=yes -C incremental=[REDACTED] -C debuginfo=2 -C link-dead-code=true -Z validate-mir -Z query-dep-graph -Z threads=4

query stack during panic:
#0 [typeck] type-checking `<impl at /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:1:1: 7:25>::{constant#0}`
#1 [used_trait_imports] finding used_trait_imports `<impl at /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:1:1: 7:25>::{constant#0}`
#2 [check_unused_traits] checking unused trait imports in crate
#3 [analysis] running analysis passes on this crate
end of query stack
error[E0425]: cannot find function, tuple struct or tuple variant `Foo` in this scope
 --> /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:4:17
  |
4 |                 Foo([0; X]);
  |                 ^^^ not found in this scope

error: aborting due to 8 previous errors

Some errors have detailed explanations: E0412, E0425, E0601, E0658, E0697.
For more information about an error, try `rustc --explain E0412`.

```

</p>
</details>

<!--
query stack:
note: compiler flags: -Z incremental-verify-ich=yes -C incremental=[REDACTED] -C debuginfo=2 -C link-dead-code=true -Z validate-mir -Z query-dep-graph -Z threads=4
#0 [typeck] type-checking `<impl at /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:1:1: 7:25>::{constant#0}`
#1 [used_trait_imports] finding used_trait_imports `<impl at /tmp/icemaker_global_tempdir.20BLeJTxFmM8/rustc_testrunner_tmpdir_reporting.ttxZjEyNw73B/mvce.rs:1:1: 7:25>::{constant#0}`
#2 [check_unused_traits] checking unused trait imports in crate
#3 [analysis] running analysis passes on this crate

-->

I guess the ICE comes from **`<const N: usize = { COMPLEX_EXPR }>`'s `[typeck]`**. See this similar ICE-case I found:

### Code

```Rust
fn foo() {}

pub struct S<const N: usize = { const { [1; foo] } }>;

fn main() {}
```

A mutant of [`tests/ui/const-generics/generic_const_exprs/inline-const-in-const-generic-defaults.rs`](https://github.com/rust-lang/rust/blob/e5177985a5493c783d6ba20960d2723d9d2e9214/tests/ui/const-generics/generic_const_exprs/inline-const-in-const-generic-defaults.rs).

It ICEs even if `foo` and `main` definitions are omitted.


<details><summary><strong>Details</strong></summary>
<p>

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.93.0-nightly (34f954f9b 2025-10-25)
binary: rustc
commit-hash: 34f954f9b7cbdb5e9b408bac1c4ff1e88b5f2719
commit-date: 2025-10-25
host: x86_64-apple-darwin
release: 1.93.0-nightly
LLVM version: 21.1.3
```

### Error output

command: `rustc`

```
error[E0308]: mismatched types
 --> 3.rs:3:45
  |
3 | pub struct S<const N: usize = { const { [1; foo] } }>;
  |                                             ^^^ expected `usize`, found fn item
  |
  = note: expected type `usize`
          found fn item `fn() {foo}`

error: internal compiler error: compiler/rustc_middle/src/ty/sty.rs:373:13: cannot find `N/#0` in param-env: ParamEnv {
                                    caller_bounds: [],
                                }


thread 'rustc' (47764347) panicked at compiler/rustc_middle/src/ty/sty.rs:373:13:
Box<dyn Any>
```

**Backtrace:**

```
thread 'rustc' (47764347) panicked at compiler/rustc_middle/src/ty/sty.rs:373:13:
Box<dyn Any>
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: <rustc_errors::diagnostic::BugAbort as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
   2: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>::{closure#0}
   3: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   4: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   5: rustc_middle::util::bug::bug_fmt
   6: <rustc_middle::ty::sty::ParamConst>::find_const_ty_from_env
   7: <rustc_trait_selection::traits::fulfill::FulfillProcessor as rustc_data_structures::obligation_forest::ObligationProcessor>::process_obligation
   8: <rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection::traits::fulfill::FulfillProcessor>
   9: <rustc_trait_selection::traits::fulfill::FulfillmentContext<rustc_trait_selection::traits::FulfillmentError> as rustc_infer::traits::engine::TraitEngine<rustc_trait_selection::traits::FulfillmentError>>::try_evaluate_obligations
  10: <rustc_hir_typeck::fn_ctxt::FnCtxt>::resolve_vars_with_obligations::<rustc_middle::ty::Ty>
  11: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  12: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_has_type_or_error::<<rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_repeat::{closure#0}>
  13: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  14: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  15: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
  16: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  17: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_const_block
  18: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  19: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
  20: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  21: rustc_hir_typeck::typeck_with_inspect::{closure#0}
  22: rustc_hir_typeck::typeck
      [... omitted 1 frame ...]
  23: rustc_mir_build::thir::pattern::check_match::check_match
      [... omitted 1 frame ...]
  24: rustc_mir_build::builder::build_mir
  25: rustc_mir_transform::mir_built
      [... omitted 1 frame ...]
  26: rustc_mir_transform::mir_const_qualif
      [... omitted 1 frame ...]
  27: rustc_mir_transform::mir_promoted
      [... omitted 2 frames ...]
  28: rustc_borrowck::mir_borrowck
      [... omitted 1 frame ...]
  29: rustc_mir_transform::mir_drops_elaborated_and_const_checked
      [... omitted 1 frame ...]
  30: rustc_mir_transform::mir_for_ctfe
      [... omitted 1 frame ...]
  31: <rustc_const_eval::const_eval::machine::CompileTimeMachine as rustc_const_eval::interpret::machine::Machine>::load_mir
  32: <rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeMachine>>::load_mir
  33: rustc_const_eval::const_eval::eval_queries::eval_to_allocation_raw_provider
      [... omitted 2 frames ...]
  34: rustc_const_eval::const_eval::valtrees::eval_to_valtree
  35: <rustc_const_eval::provide::{closure#0} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, rustc_middle::ty::PseudoCanonicalInput<rustc_middle::mir::interpret::GlobalId>)>>::call_once
      [... omitted 2 frames ...]
  36: <rustc_middle::ty::context::TyCtxt>::const_eval_global_id_for_typeck
  37: <rustc_middle::ty::context::TyCtxt>::const_eval_resolve_for_typeck
  38: rustc_trait_selection::traits::try_evaluate_const
  39: <rustc_trait_selection::traits::normalize::AssocTypeNormalizer as rustc_type_ir::fold::TypeFolder<rustc_middle::ty::context::TyCtxt>>::fold_const
  40: <rustc_trait_selection::traits::normalize::AssocTypeNormalizer as rustc_type_ir::fold::TypeFolder<rustc_middle::ty::context::TyCtxt>>::fold_predicate
  41: rustc_trait_selection::traits::normalize::normalize_with_depth_to::<rustc_middle::ty::predicate::Clause>::{closure#0}
  42: <rustc_trait_selection::traits::engine::ObligationCtxt<rustc_trait_selection::traits::FulfillmentError>>::normalize::<rustc_middle::ty::predicate::Clause>
  43: <rustc_hir_analysis::check::wfcheck::WfCheckingCtxt>::normalize::<rustc_middle::ty::predicate::Clause>
  44: <core::iter::adapters::chain::Chain<core::iter::adapters::flatten::FlatMap<core::iter::adapters::zip::Zip<alloc::vec::into_iter::IntoIter<rustc_middle::ty::predicate::Clause>, alloc::vec::into_iter::IntoIter<rustc_span::span_encoding::Span>>, thin_vec::ThinVec<rustc_infer::traits::Obligation<rustc_middle::ty::predicate::Predicate>>, rustc_hir_analysis::check::wfcheck::check_where_clauses::{closure#4}>, core::iter::adapters::map::Map<core::iter::adapters::flatten::FlatMap<core::slice::iter::Iter<(rustc_middle::ty::predicate::Clause, rustc_span::span_encoding::Span)>, core::option::Option<(rustc_middle::ty::predicate::Clause, rustc_span::span_encoding::Span)>, rustc_hir_analysis::check::wfcheck::check_where_clauses::{closure#2}>, rustc_hir_analysis::check::wfcheck::check_where_clauses::{closure#3}>> as core::iter::traits::iterator::Iterator>::next
  45: rustc_hir_analysis::check::wfcheck::check_where_clauses
  46: rustc_hir_analysis::check::wfcheck::check_type_defn
  47: rustc_hir_analysis::check::check::check_item_type
  48: rustc_hir_analysis::check::wfcheck::check_well_formed
      [... omitted 1 frame ...]
  49: rustc_hir_analysis::check::wfcheck::check_type_wf
      [... omitted 1 frame ...]
  50: rustc_hir_analysis::check_crate
  51: rustc_interface::passes::analysis
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/Volumes/T7/workspace/scratch_250918/rustc-ice-2025-10-26T06_23_02-37201.txt` to your bug report

query stack during panic:
#0 [typeck] type-checking `S::{constant#0}`
#1 [check_match] match-checking `S::{constant#0}`
#2 [mir_built] building MIR for `S::{constant#0}`
#3 [mir_const_qualif] const checking `S::{constant#0}`
#4 [mir_promoted] promoting constants in MIR for `S::{constant#0}`
#5 [mir_borrowck] borrow-checking `S::{constant#0}`
#6 [mir_drops_elaborated_and_const_checked] elaborating drops for `S::{constant#0}`
#7 [mir_for_ctfe] caching mir of `S::{constant#0}` for CTFE
#8 [eval_to_allocation_raw] const-evaluating + checking `S::{constant#0}`
#9 [eval_to_valtree] evaluating type-level constant
#10 [check_well_formed] checking that `S` is well-formed
#11 [check_type_wf] checking that types are well-formed
#12 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
```

### Regression in 

`Regression in nightly-2024-06-07`

(searched toolchains nightly-2024-01-01 through nightly-2025-10-25)


### Notes
- Following cases ICE too:
  - `impl<const N: usize = { const { [1; foo] } }> for PartialEq for True {}`
  - `struct S<const N: usize = { static || Foo([0; X]) }>;`
- I was unsure whether to open a new issue for this ICE. The query stacks differ but are similar (`analysis-trait_check-typeck` vs `analysis-MIR_building-typeck`). Any guidance?

#### Related ICEs to consider:
ICE issues with the location `compiler/rustc_middle/src/ty/sty.rs Line-373`
- issue-146384 (requires generic_const_exprs)
- issue-143117 (requires generic_const_exprs)

</p>
</details>

--- a/tests/ui/associated-consts/assoc-const-equality-region-var-hash-ice.rs
+++ b/tests/ui/associated-consts/assoc-const-equality-region-var-hash-ice.rs
@@ -1,0 +1,17 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/148620>.
+
+trait Trait<'a> {
+    const K: &'a ();
+}
+
+fn foo() -> dyn Trait<'r, K = {}> {
+    //~^ ERROR use of undeclared lifetime name `'r`
+    //~| ERROR associated const equality is incomplete
+    {
+        6
+        //~^ ERROR mismatched types
+    }
+    5
+}
+
+fn main() {}

--- a/tests/ui/associated-consts/assoc-const-equality-region-var-hash-ice.stderr
+++ b/tests/ui/associated-consts/assoc-const-equality-region-var-hash-ice.stderr
@@ -1,0 +1,41 @@
+error[E0261]: use of undeclared lifetime name `'r`
+  --> $DIR/assoc-const-equality-region-var-hash-ice.rs:7:23
+   |
+LL | fn foo() -> dyn Trait<'r, K = {}> {
+   |                       ^^ undeclared lifetime
+   |
+   = note: for more information on higher-ranked polymorphism, visit https://doc.rust-lang.org/nomicon/hrtb.html
+help: consider making the bound lifetime-generic with a new `'r` lifetime
+   |
+LL | fn foo() -> dyn for<'r> Trait<'r, K = {}> {
+   |                 +++++++
+help: consider introducing lifetime `'r` here
+   |
+LL | fn foo<'r>() -> dyn Trait<'r, K = {}> {
+   |       ++++
+
+error[E0658]: associated const equality is incomplete
+  --> $DIR/assoc-const-equality-region-var-hash-ice.rs:7:27
+   |
+LL | fn foo() -> dyn Trait<'r, K = {}> {
+   |                           ^^^^^^
+   |
+   = note: see issue #132980 <https://github.com/rust-lang/rust/issues/132980> for more information
+   = help: add `#![feature(min_generic_const_args)]` to the crate attributes to enable
+   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
+
+error[E0308]: mismatched types
+  --> $DIR/assoc-const-equality-region-var-hash-ice.rs:11:9
+   |
+LL |         6
+   |         ^ expected `()`, found integer
+   |
+help: you might have meant to return this value
+   |
+LL |         return 6;
+   |         ++++++  +
+
+error: aborting due to 3 previous errors
+
+Some errors have detailed explanations: E0261, E0308, E0658.
+For more information about an error, try `rustc --explain E0261`.

--- a/tests/ui/const-generics/defaults/const-param-default-body-param-env-ice.rs
+++ b/tests/ui/const-generics/defaults/const-param-default-body-param-env-ice.rs
@@ -1,0 +1,15 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/148096>.
+
+impl<const N: usize = { || [0; X] }, X> dyn PartialEq<X> {}
+//~^ ERROR generic parameters with a default must be trailing
+//~| ERROR generic parameter defaults cannot reference parameters before they are declared
+//~| ERROR defaults for generic parameters are not allowed here
+//~| ERROR the const parameter `N` is not constrained by the impl trait, self type, or predicates
+//~| ERROR cannot define inherent `impl` for a type outside of the crate where the type is defined
+
+fn foo() {}
+
+pub struct S<const N: usize = { const { [1; foo] } }>;
+//~^ ERROR mismatched types
+
+fn main() {}

--- a/tests/ui/const-generics/defaults/const-param-default-body-param-env-ice.stderr
+++ b/tests/ui/const-generics/defaults/const-param-default-body-param-env-ice.stderr
@@ -1,0 +1,50 @@
+error: generic parameters with a default must be trailing
+  --> $DIR/const-param-default-body-param-env-ice.rs:3:12
+   |
+LL | impl<const N: usize = { || [0; X] }, X> dyn PartialEq<X> {}
+   |            ^
+
+error[E0128]: generic parameter defaults cannot reference parameters before they are declared
+  --> $DIR/const-param-default-body-param-env-ice.rs:3:32
+   |
+LL | impl<const N: usize = { || [0; X] }, X> dyn PartialEq<X> {}
+   |                                ^ cannot reference `X` before it is declared
+
+error: defaults for generic parameters are not allowed here
+  --> $DIR/const-param-default-body-param-env-ice.rs:3:6
+   |
+LL | impl<const N: usize = { || [0; X] }, X> dyn PartialEq<X> {}
+   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error[E0207]: the const parameter `N` is not constrained by the impl trait, self type, or predicates
+  --> $DIR/const-param-default-body-param-env-ice.rs:3:6
+   |
+LL | impl<const N: usize = { || [0; X] }, X> dyn PartialEq<X> {}
+   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unconstrained const parameter
+   |
+   = note: expressions using a const parameter must map each value to a distinct output value
+   = note: proving the result of expressions other than the parameter are unique is not supported
+
+error[E0308]: mismatched types
+  --> $DIR/const-param-default-body-param-env-ice.rs:12:45
+   |
+LL | pub struct S<const N: usize = { const { [1; foo] } }>;
+   |                                             ^^^ expected `usize`, found fn item
+   |
+   = note: expected type `usize`
+           found fn item `fn() {foo}`
+   = note: array length can only be `usize`
+
+error[E0116]: cannot define inherent `impl` for a type outside of the crate where the type is defined
+  --> $DIR/const-param-default-body-param-env-ice.rs:3:1
+   |
+LL | impl<const N: usize = { || [0; X] }, X> dyn PartialEq<X> {}
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ impl for type defined outside of crate
+   |
+   = help: consider defining a trait and implementing it for the type or using a newtype wrapper like `struct MyType(ExternalType);` and implement it
+   = note: for more details about the orphan rules, see <https://doc.rust-lang.org/reference/items/implementations.html?highlight=orphan#orphan-rules>
+
+error: aborting due to 6 previous errors
+
+Some errors have detailed explanations: E0116, E0128, E0207, E0308.
+For more information about an error, try `rustc --explain E0116`.

